### DISCUSSION TITLE
Release version v.1.0.0-beta

### DISCUSSION THREAD

**Discussion by FriendsOfGalaxy** [FriendsOfGalaxy/galaxy-integration-steam#131] [Show and tell]
## Release version v.1.0.0-beta
I would like to share with the whole GOG Galaxy 2.0 Steam integration community a new beta version
of the Steam plugin:
- [windows.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/7129605/windows.zip)
- [macos.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/7129606/macos.zip)

The main feature of this soft release is a multi-backend solution to deal with
a disconnecting websocket connection with Steam (`eresult: 5` error). The solution consists two backend types.

1. The classic `Steam Network`
    - uses internal Steam protocols 
    - supports the following features: library, game times, achievements, importing tags, friends presence
2. `Public Profiles`
    - works without the need to provide user credentials
    - requires the user's steamcommunity.com profile to be set to public, with access to the games library at the very least
    - supports the following features: library, game times, achievements, importing tags

The initial `backend` (with the login window) is set to use `Steam Network` and to switch to `Public Profiles` automatically 
when the stored token becomes invalid (for the main reason see #74).

NOTE: the way the plugin interacts with local games and local Steam client is the same for all `backends`.

## Installation
* make sure Galaxy is closed
* remove the currently installed plugin directory (if present), usually<br>
`%localappdata%\GOG.com\Galaxy\plugins\installed\steam_ca27391f-2675-49b1-92c0-896d43afa4f8`
* create a new folder under a name of your choice (the name doesn't matter) at the following path:<br>
`%localappdata%\GOG.com\Galaxy\plugins\installed\`
* copy the custom build files to the newly created folder

## Config file location
Above described behavior is configurable by `config file`, which will be generated after first connection of this plugin.

- Windows:
`%localappdata%\GOG.com\Galaxy\plugins\installed\steam_plugin_config.ini`

- MacOS:
`~/Library/Application Support/GOG.com/Galaxy/plugins/installed/steam_plugin_config.ini`

**Comment by ReiTung** [comment]
Hi, 

wow, many thanks for that, your efforts are truly appreciated.

While I will not have too much time to play and therefore test deeply over the next weeks, I definitely will start working with that version if using Galaxy. Hence, a few questions to better understand and adopt my testing accordingly.

- I do understand that this is currently a Beta Version only, so is currently being a separate branch beside v0.60?
- having said that, I would understand the version is not automatically updating, i.e. v0.60 will not be overwritten with v1.0.0. Respectively, v1.0.0 would have to be manually installed using the steps below
- would v1.0.0 update itself then if newer versions of that branch are released, i.e. does the beta already have an autoupdate function?
- having said that, my assumption is that in case of serious issues, a fallback to v0.60+ would easily be possible by deleting the v1.0.0 sources and recoping the v0.60 code
- will v0.60 still be maintained/updated in case of issues reported?

Apologies if the following questions regarding functionality of v1.0.0 sounds a bit silly, but for my testing might be important to understand:

- do I get it right that the two different backends have been introduced because the real issue (WHY THE HECK THE TOKEN IS NOT RFRESHED :)) is still not identified/resolved?
- i.e. does it mean that while the authentication will start with full Steam functionality the token will still become invalid (err 5) after some time, but the plugin will then try to switch to respective public profile then to mitigate (instead of refreshing the token)?
- will the plugin try to refresh the token again later on, i.e. fall back to full functionality again on its own or would it require a reconnect?
- in case v1.0.0 fails for whatever reason, is there a similar "reconnect" functionality *without* deleting Steam data in the Galaxy Library (i.e. no loss of the complete historic data) similar to what we have in the latest versions of "old" plugin?

and finally, any specific scenarios you want to have tested? I assume logfiles can be submitted as usual and will log which version of the plugin is used?

thanks again for all your efforts!

Rei Tung

**Reply by FriendsOfGalaxy** [reply]
Firstly, thank you @ReiTung for kind words! Now I'll try to answer all of your questions :)


> I do understand that this is currently a Beta Version only, so is currently being a separate branch beside v0.60?

Yes, for now it is beta version only, branched from `v0.60`, but soon I am going to release it fully.

***

> having said that, I would understand the version is not automatically updating, i.e. v0.60 will not be overwritten with v1.0.0. Respectively, v1.0.0 would have to be manually installed using the steps below

Indeed, for now version `v0.60` is the latest released version which Galaxy knows, that's why this beta has to be installed manually.

***

> would v1.0.0 update itself then if newer versions of that branch are released, i.e. does the beta already have an autoupdate function?

No, beta version doesn't have an autoupdate function, so if there are any fixes/updates I will also release it in this Github discussion.

***

>having said that, my assumption is that in case of serious issues, a fallback to v0.60+ would easily be possible by deleting the v1.0.0 sources and recoping the v0.60 code

In case of any serious issues/problems with this beta release you can always remove it from `installed` directory and let Galaxy install official release v0.60

***

>will v0.60 still be maintained/updated in case of issues reported?

Before production release of beta, `v0.60` is still main and stable version of Steam plugin, and it is maintained. But version `v1.0.0` contains whole solution from `v0.60` as a primary `backend` then if reported issue will be related only with `v0.60` then for now whole fix will be in both versions.

***

>do I get it right that the two different backends have been introduced because the real issue (WHY THE HECK THE TOKEN IS NOT RFRESHED :)) is still not identified/resolved?

Partially yes, because that's why there exists some backend switching mechanism. But the main reason is that some users just don't want to be logged in to Steam, they don't care about any async features, they just want to see their game library (and eventually achievements) to play. Public profile backend let users to do it :)

***

>i.e. does it mean that while the authentication will start with full Steam functionality the token will still become invalid (err 5) after some time, but the plugin will then try to switch to respective public profile then to mitigate (instead of refreshing the token)?

If during authentication Steam response will be `err 5` then instead of crash, like current version do, plugin will try to switch to public profile (try because it depends on your profile privacy). If someone doesn't want this behaviour there is possibility to disable it by setting fallback backend to `steam_network` in config file.

***

>will the plugin try to refresh the token again later on, i.e. fall back to full functionality again on its own or would it require a reconnect?

Plugin behaves the same on every restart of Galaxy. If you'll get `err 5` and you'll switch to public profiles, then on next Galaxy restart this behaviour will be the same. But as far as I know Steam behaviour, if token is invalid once, then it will invalid on every restart.

***

>in case v1.0.0 fails for whatever reason, is there a similar "reconnect" functionality without deleting Steam data in the Galaxy Library (i.e. no loss of the complete historic data) similar to what we have in the latest versions of "old" plugin?

It depends on which backend you are. For `steam_network` it works as in "old" plugin.

***

About scenarios, it would be cool to test it exactly the same as non-dev user do. Then, log in with public and private profile and check how plugin works, let check changing configuration in config file. For me, the most important thing is no regression. 

Thank you for all your questions and don't hesitate to ask next!

**Comment by ReiTung** [comment]
thanks, all clear so far. I will install today and use it over the next days in case I actually find some time to play :)

one additional question though: is there some obvious way to identify in which backend mode the Plugin is at a given time? and additionally, a way to be "informed" when the plugin falls back to profiles only?
Maybe it would be cool to have this information shown near the "Connected" status in the plugin status, something like "Connected (Steam backend)" vs. "Connected (Profile only)".

Thinking of that because once the plugin has fallen back to Profiles, it will not switch back to Steam Backend if I understand correctly ... so people will recognize only by accident, either if no more data synced (if Profile not set to public) or missing sync of friend list

**Reply by FriendsOfGalaxy** [reply]
Oh yes, I didn't mention it. When plugin switch to public profile mode, username in Galaxy -> Settings -> Steam plugin row will have `(public)` suffix :)

**Comment by ReiTung** [comment]
Just as a quick headsup - no regressions found for the known "standard scenarios" so far (full Steam connection). Installation OK, logging OK, logon OK. Synch and asynch import of Steam data OK so far - looks good.
However, due to time constraints, none of the more "complicated" scenarios (multiple PCs on same network, parallel execution of Steam etc.) tried so far.

**Reply by FriendsOfGalaxy** [reply]
However thank you for all tests!
